EngineLayer/Calibration/CalibrationEngine.cs

```diff
@@ -1,332 +1,249 @@
-                      }
-                      else
-                      {
-                          Func<MzPeak, double> theFunc = x => x.Mz - ms1predictor.Predict(new double[] { x.Mz, scan.RetentionTime, Math.Log(scan.TotalIonCurrent), scan.InjectionTime.HasValue ? Math.Log(scan.InjectionTime.Value) : double.NaN, Math.Log(x.Intensity) });
-                          scan.MassSpectrum.ReplaceXbyApplyingFunction(theFunc);
-                      }
-                  }
-              }
-              );
+            return averageRelativeErrors;
         }
 
-        private RegressionForestModel GetRandomForestModel(List<(double[] xValues, double yValue)> myInputs, double fracForTraining)
+        private double CalculateAverageRelativeErrors(List<(double massError, double logIntensity)> localRelativeErrors)
         {
-            // create a machine learner
-            var learner = new RegressionRandomForestLearner();
-            var metric = new MeanAbsolutErrorRegressionMetric();
-
-            var splitter = new RandomTrainingTestIndexSplitter<double>(trainingPercentage: fracForTraining, seed: RandomSeed);
+            double logIntensityToSubtract = localRelativeErrors.Min(x => x.logIntensity) - 1; // normalize each log intensity so that the minimum log intensity is 1. 
+            //This is to more heavily weight high log intensities, so that a difference like 20 and 22 is now 1 and 3. the "-1" prevents zero values
+            //The maximum possible difference in weight becomes ~1:10, such that low intensity points are still considered
+            double weightedSumOfErrors = localRelativeErrors.Sum(x => x.massError * (x.logIntensity - logIntensityToSubtract));
+            double sumOfIntensities = localRelativeErrors.Sum(x => x.logIntensity - logIntensityToSubtract);
+            return weightedSumOfErrors / sumOfIntensities;
+        }
 
-            // put x values into a matrix and y values into a 1D array
-            var myXValueMatrix = new F64Matrix(myInputs.Count, myInputs.First().xValues.Length);
-            for (int i = 0; i < myInputs.Count; i++)
+        private double[] SmoothErrors(double[] relativeErrors, int numberOfScansUsedForSmoothingOnEachSide)
+        {
+            //impute missing values
+            //not all scans are guarenteed to contain data points. We can infer these data point with nearby points.
+            for (int index = 0; index < relativeErrors.Length; index++)
             {
-                for (int j = 0; j < myInputs.First().xValues.Length; j++)
+                if (relativeErrors[index] == 0) //if there were no points, then the value should be perfectly zero (the double default)
                 {
-                    myXValueMatrix[i, j] = myInputs[i].xValues[j];
+                    int startingBlankIndex = index;
+                    for (; index < relativeErrors.Length && relativeErrors[index] == 0; index++) { };
```

Comment:
This is pretty opaque. Any way you can do the math without a loop?

Reply:
I need to figure out the next index at which there is a datapoint, so it has to loop through until it hit  a scan with a data point... I could comment it?

Reply:
I think a while loop would be clearer. These empty `for` loops without an initializer statement are really strange, and they're effectively while loops.

`while (index < relativeErrors.Length && relativeErrors[index] == 0) { index++; }`